From bebcb1c4303c3c1bfe4dd9051593a56b2565e0a4 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Thu, 16 May 2013 16:44:09 -0700
Subject: [PATCH 0690/1532] syslib: add a workaround to set fmn pcim base for
 xlp9xx

  o The workaround should be removed once uboot for xlp9xx is working
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 arch/mips/netlogic/lib/syslib/src/nlm_hal.c | 25 +++++++++++++++++++++----
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 1551efe..254761d 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -400,13 +400,30 @@ __inline__ void nlm_hal_init(void)
 		xlp_io_base = KSEG1 + 0x18000000;
 
         	/* PCI enumeration of supported devices*/
-		xlp_fmn_base[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_FMN)), PCI_MEM_BAR_0);
 
-		xlp_mac_base[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_NAE)), PCI_MEM_BAR_0); //0x18018000
+		/* FIXME: this is a workaround for simulator, once uboot is ready for simulator,
+		 * the below should be removed.
+		 */
+		if (is_nlm_xlp9xx()) {
+			uint32_t fmn_bar = 0xd0000000;
+
+			nlm_hal_write_32bit_reg((0x18000000 | (1 << 20) | XLP_CFG_BASE(0, XLP_FMN)), 0x4, fmn_bar);
+			nlm_hal_write_32bit_reg((0x18000000 | (1 << 20) | XLP_CFG_BASE(0, XLP_FMN)), 0x5, 0x0);
+			xlp_fmn_base[0] = 0x9000000000000000ULL | fmn_bar;
+
+			xlp_mac_base[0] = 0x0;
+			xlp_nae_base[0] = 0x0;
+			xlp_nae_base[0] = 0x0;
+		} else {
+			xlp_fmn_base[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_FMN)), PCI_MEM_BAR_0);
+
+			xlp_mac_base[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_NAE)), PCI_MEM_BAR_0); //0x18018000
 		//printf("Node:%d NAE_MAC_Base:%lX\n", node, xlp_mac_base[node]);
-		xlp_nae_base[node] = xlp_mac_base[node] + 0xe000;
+			xlp_nae_base[node] = xlp_mac_base[node] + 0xe000;
+
+			xlp_poe_base_pcim[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_POE)), PCI_MEM_BAR_0);	//0x18019000
+		}
 
-		xlp_poe_base_pcim[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_POE)), PCI_MEM_BAR_0);	//0x18019000
 		xlp_poe_base_pcie[node] = (xlp_io_base | XLP_CFG_BASE(node, XLP_POE)) & 0x1fffffff; /* For now . Will be fixed soon.*/
 
 		xlp_sys_base[node] = (xlp_io_base | XLP_CFG_BASE(node, XLP_SYS)) & 0x1fffffff; /*For now . Will be fixed soon.*/
-- 
1.9.1

